enclave-runtime/src/initialization/parentchain/parachain.rs

```diff
@@ -0,0 +1,115 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+use crate::{
+	error::Result,
+	initialization::{
+		global_components::{
+			EnclaveExtrinsicsFactory, EnclaveNodeMetadataRepository, EnclaveOCallApi,
+			EnclaveParentchainBlockImportDispatcher, EnclaveStfExecutor, EnclaveValidatorAccessor,
+			GLOBAL_FULL_PARACHAIN_HANDLER_COMPONENT, GLOBAL_OCALL_API_COMPONENT,
+			GLOBAL_STATE_HANDLER_COMPONENT,
+		},
+		parentchain::common::{
+			create_extrinsics_factory, create_offchain_immediate_import_dispatcher,
+			create_parentchain_block_importer, create_sidechain_triggered_import_dispatcher,
+			HeaderFor,
+		},
+	},
+};
+use codec::{Decode, Encode};
+use itc_parentchain::light_client::{
+	concurrent_access::ValidatorAccess, light_client_init_params::ParachainParams, LightClientState,
+};
+use itp_component_container::{ComponentGetter, ComponentInitializer};
+use itp_settings::worker_mode::{ProvideWorkerMode, WorkerMode};
+use itp_types::Block as ParentchainBlock;
+use std::{sync::Arc, vec::Vec};
+
+#[derive(Clone)]
+pub struct FullParachainHandler {
+	pub genesis_header: HeaderFor<ParentchainBlock>,
+	pub node_metadata_repository: Arc<EnclaveNodeMetadataRepository>,
+	pub stf_executor: Arc<EnclaveStfExecutor>,
+	pub validator_accessor: Arc<EnclaveValidatorAccessor>,
+	pub extrinsics_factory: Arc<EnclaveExtrinsicsFactory>,
+	pub import_dispatcher: Option<Arc<EnclaveParentchainBlockImportDispatcher>>,
+}
+
+impl FullParachainHandler {
+	pub fn init<WorkerModeProvider: ProvideWorkerMode>(encoded_params: Vec<u8>) -> Result<Vec<u8>> {
```

Comment:
The idea behind this is that only the parentchain specific initialization part knows what kind of params it should expect. 

E.g. if we decode the params before this part:
```rust
let params = ParachainParams::<HeaderFor<ParentchainBlock>>::decode(&mut encoded_params.as_slice())?;
```
We need to define the header type of the parentchain somewhere before calling this init function.

Reply:
Hmm, I can only partly understand. :) Why do we not know about the parentchain block from the outside? I would argue that the outside does know, the header, otherwise, why would it know that the `FullParentchainBlockHandler` needs to be initialized?

Even if the outside does not know about the specific type, we enforce the type, when passing the header in.

```rust
struct FullParachainHandler<ParentchainBlock: ParentchainBlockTrait> {
     // stuff
}

impl<ParentchainBlock: ParentchainBlockTrait> FullParentchainHandler<ParentchainBlock> {
      pub fn init<WProvidver, ParentchainBlock>(params: ParenchtChanBlockParams<HeaderFor<ParentchainBlock>) -> Result<HeaderFor<ParentchainBlock>> {
   // init stuff
      }
}
```

You can then call the code from the outside like:

```rust
// this works also if the full parentchain handler is not generic.
FullParachainHandler::<ParentchainBlock>(Decode::decode(encoded_params)?);
```

I don't have a strong opinion here (but my approach would give a few more compile time guarantees), but I am curious about your answer as well for my own understanding. I need to know the low level details again as of now :)

Reply:
Yeah, you're right. Sorry, I had a block there somehow 🙈 . Going to fix it right away.

Reply:
Fixed in [38b8b9f](https://github.com/integritee-network/worker/pull/1078/commits/38b8b9fe5389e4afa4bd0a03adb3a4c0f7bd554b)

Does that work for you? :)